File: apps/web/lib/localization/postal.ts
import { normalizeDigits } from './normalize';

/**
 * Validates whether the given string is a valid 10-digit Iranian postal code.
 */
export function validatePostalCode(str: string): { isValid: boolean; normalized?: string; error?: string } {
  if (!str) {
    return { isValid: false, error: 'کد پستی وارد نشده است.' };
  }

  const normalized = normalizeDigits(str).trim().replace(/[\s-]/g, '');

  if (!/^\d{10}$/.test(normalized)) {
    return { isValid: false, error: 'کد پستی باید دقیقاً یک عدد ۱۰ رقمی باشد.' };
  }

  return { isValid: true, normalized };
}

File: apps/api/internal/localization/postal.go
package localization

import (
	"errors"
	"regexp"
	"strings"
)

var (
	ErrInvalidPostalCode = errors.New("کد پستی باید یک عدد ۱۰ رقمی معتبر باشد")
	postalCodeRegex      = regexp.MustCompile(`^\d{10}$`)
)

// ValidatePostalCode normalizes digits and validates that the string is exactly a 10-digit Iranian postal code.
func ValidatePostalCode(postalCode string) (string, error) {
	normalized := NormalizeDigits(strings.TrimSpace(postalCode))
	normalized = strings.ReplaceAll(normalized, "-", "")
	normalized = strings.ReplaceAll(normalized, " ", "")

	if !postalCodeRegex.MatchString(normalized) {
		return "", ErrInvalidPostalCode
	}
	return normalized, nil
}
